FLT/Deformations/RepresentationTheory/Basic.lean

```diff
@@ -0,0 +1,379 @@
+    (ρ.conj e).IsUnramifiedAt v where
+  localInertiaGroup_le := (GaloisRep.IsUnramifiedAt.localInertiaGroup_le (ρ := ρ)).trans (by simp)
+
+instance [IsTopologicalRing B] [Algebra A B] [ContinuousSMul A B]
+    [Module.Finite A M] [Module.Free A M] (ρ : GaloisRep K A M) (v : Ω K) [ρ.IsUnramifiedAt v] :
+    (ρ.baseChange B).IsUnramifiedAt v :=
+  ⟨(GaloisRep.IsUnramifiedAt.localInertiaGroup_le (ρ := ρ)).trans
+    (((ρ.adic v).ker_baseChange (B := B)))⟩
+
+variable [Module.Free A M] [Module.Finite A M] [Module.Free A N] [Module.Finite A N]
+
+/-- The character polynomial of the frobenious conjugacy class at `v` under `ρ`. -/
+noncomputable
+def GaloisRep.charFrob (ρ : GaloisRep K A M) : Polynomial A := (ρ.adic v Frobᵥ).charpoly
+
+omit [IsTopologicalRing A] in
+lemma GaloisRep.charFrob_eq (ρ : GaloisRep K A M) [ρ.IsUnramifiedAt v] (σ : Γ Kᵥ)
+    (hσ : IsArithFrobAt 𝒪ᵥ σ (𝔪 (IntegralClosure 𝒪ᵥ (Kᵥᵃˡᵍ)))) :
+    (ρ.adic v σ).charpoly = ρ.charFrob v := by
+  have := IsUnramifiedAt.localInertiaGroup_le (ρ := ρ)
+    (hσ.mul_inv_mem_inertia (Field.AbsoluteGaloisGroup.isArithFrobAt_adicArithFrob v))
+  replace this := congr($this * ρ.adic v Frobᵥ)
+  simp only [ContinuousMonoidHom.coe_toMonoidHom, ← map_mul, MonoidHom.coe_coe, one_mul,
+    inv_mul_cancel_right] at this
+  rw [this, charFrob]
+
+section Flat
+
+set_option linter.unusedVariables false in
+/-- The underlying space of a galois rep. This is a type class synonym that allows `G` to act
+on it via `ρ`. -/
+@[nolint unusedArguments]
+abbrev GaloisRep.Space (ρ : GaloisRep K A M) : Type _ := M
+
+instance (ρ : GaloisRep K A M) : DistribMulAction (Γ K) ρ.Space where
+  smul g v := ρ g v
+  one_smul := by simp [instHSMul]
+  mul_smul := by simp [instHSMul]
+  smul_zero := by simp [instHSMul]
+  smul_add := by simp [instHSMul]
+
+attribute [instance 10000]
+  MulSemiringAction.toDistribMulAction
+  DistribMulAction.toMulAction
+  MulAction.toSMul
+
+open TensorProduct in
+/-- A galois rep `ρ : Γ K → Aut_A(M)` has a flat prolongation at `v` if `M` (when viewed as a
+`Γ Kᵥ`) module is isomorphic to the geometric points of a finite etale hopf algebra over `Kᵥ`, and
+there exists an finite flat hopf algebra over `𝒪ᵥ` whose generic fiber is isomorphic to it.
+In particular this requires `M` (and by extension `A`) to have finite cardinality. -/
+def GaloisRep.HasFlatProlongationAt (ρ : GaloisRep K A M) : Prop :=
+  ∃ (G : Type uK) (_ : CommRing G) (_ : HopfAlgebra 𝒪ᵥ G)
+    (_ : Module.Flat 𝒪ᵥ G) (_ : Module.Finite 𝒪ᵥ G) (_ : Algebra.Etale Kᵥ (Kᵥ ⊗[𝒪ᵥ] G))
+    (f : Additive (Kᵥ ⊗[𝒪ᵥ] G →ₐ[Kᵥ] Kᵥᵃˡᵍ) →+[Γ Kᵥ] (ρ.adic v).Space),
+    Function.Bijective f
+
+/-- A galois rep `ρ : Γ K → Aut_A(M)` is flat at `v` if `A/mⁿ ⊗ M` has a flat prolongation at `v`
+for all `n`. -/
+class GaloisRep.IsFlatAt [IsLocalRing A] (ρ : GaloisRep K A M) : Prop where
+  cond : ∀ n : ℕ, n ≠ 0 →
+    (ρ.baseChange (A ⧸ IsLocalRing.maximalIdeal A ^ n)).HasFlatProlongationAt v
```

Comment:
I guess it would be better if these were predicates on reps of local Galois groups but right now we still have no way to say "local field" (and in fact the correct context is something like "rep of absolute Galois group of finite extension of field of fractions of Witt vectors of perfect char p field")